[클린코드 7기 김진주] 로또 미션 STEP 1

devpearlkim opened the pull request:
안녕하세요 리뷰어님,
리뷰 감사드립니다.
아래와 같은 부분을 고민하고 있는데 의견 주시면 감사하겠습니다.

## 고민한 부분 & 궁금한 점

### LottoService의 setWinningNumbers() 호출 방식
현재 main()에서는 LottoService를 초기화할 때 amount 값을 전달하여 로또 티켓을 먼저 생성한 후,
당첨 번호를 입력받아 setWinningNumbers()를 호출하는 구조입니다.

게임의 핵심 데이터인 당첨 번호를 중간에 변경하는 방식이 적절한지,
혹은 더 나은 접근 방식이 있을지 고민됩니다.

### LottoResult의 역할 분리
현재 LottoResult가
- 당첨 결과를 계산하는 역할
-  입력받은 당첨 번호의 검증(Validation) 역할

을 함께 수행하고 있습니다.
이러한 구조가 괜찮은지, 혹은 검증 로직을 분리하는 것이 더 나은 방향일지 궁금합니다.

devpearlkim committed:
feat: Implement Lotto purchase logic with input validation and ticket calculation

- Added validation to ensure purchase amount is greater than 0 and a multiple of 1,000
- Implemented logic to calculate the number of Lotto tickets based on the purchase amount.
- Created tests for various purchase amounts, including valid and invalid cases.

devpearlkim committed:
feat: Implement lotto number generation with random unique numbers

- Generate lotto numbers between 1 and 45
- Ensure each ticket contains 6 unique numbers
- Validate correct ticket generation based on purchase amount
- Add test cases for number range, uniqueness, and ticket differentiation

devpearlkim committed:
feat: Add LottoResult class with validation for winning numbers and bonus number

- Added validation for positive integers, length of winning numbers, duplicates, and bonus number uniqueness.
- Integrated custom error messages for each validation rule.
- Wrote test cases to validate LottoResult class functionality, including edge cases for invalid input.
- Implemented UI logic in input.js to handle user input for purchase amount, winning numbers, and bonus number.
- Implemented output.js to display generated lotto tickets and results.

devpearlkim committed:
feat: Develop countMatchingNumber and calculateRank methods with corresponding tests

- Introduced countMatchingNumber method to count matching numbers between ticket and winning numbers.
- Developed calculateRank method to determine the prize based on matching numbers and bonus number.
- Wrote unit tests for both methods, covering edge cases like matching numbers, duplicate numbers, and ticket validity.

devpearlkim committed:
feat: Display winning statistics and refactor LottoService

- Add a function to display winning statistics in the correct order
- Refactor LottoService to separate winning number assignment
- Move winning number logic from constructor to setWinningNumbers()

devpearlkim committed:
feat: Implement profit rate calculation in LottoStatistics

devpearlkim committed:
fix: Validate winning numbers to be between 1 and 45

devpearlkim committed:
chore: Update entry points and npm scripts for step-console and step-web

devpearlkim committed:
chore: Optimize lotto number generation using Fisher-Yates Shuffle

devpearlkim committed:
fix: Adjust profit rate calcuration to support negative values

Tanney-102 commented on __tests__/LottoResult.test.js:
```diff
@@ -0,0 +1,120 @@
+describe("LottoResult", () => {
+  describe("Number validation", () => {
+    it("should throw an error if winning numbers contain numbers outside the range 1-45", () => {
+      expect(() => new LottoResult([-1, 2, 3, 4, 5, 6], 7)).toThrow(
+        ERROR_MESSAGES.WINNING_NUMBERS_INVALID,
+      );
+      expect(() => new LottoResult([1, 2, 3, 4, 5, 46], 7)).toThrow(
+        ERROR_MESSAGES.WINNING_NUMBERS_INVALID,
+      );
+      expect(() => new LottoResult([0, 2, 3, 4, 5, 6], 7)).toThrow(
+        ERROR_MESSAGES.WINNING_NUMBERS_INVALID,
+      );
+    });
```
경계값만 잘 테스트 해줘도 좋을 듯 해요. lower bound 에 대해서 -1, 0 둘 다 테스트 할 필요는 없어보입니다.

Tanney-102 commented on src/view/ui/input.js:
```diff
@@ -0,0 +1,32 @@
+import readline from "readline";
+
+function readLineAsync(query) {
+  return new Promise((resolve, reject) => {
+    const rl = readline.createInterface({
+      input: process.stdin,
+      output: process.stdout,
+    });
+
+    rl.question(query, (input) => {
+      rl.close();
+      resolve(input.trim());
+    });
+  });
+}
+
+export async function getPurchaseAmount() {
+  const input = await readLineAsync("구입금액을 입력해 주세요: ");
+  return parseInt(input, 10);
+}
```
공용으로 사용할 수 있는 UI 함수들과 로또 앱에서만 사용할 UI 함수를 별도의 모듈로 관리하면 공통 로직의 재사용성을 확보할 수 있을 것 같아요!

Tanney-102 commented on src/validation.js:
```diff
@@ -0,0 +1,11 @@
+export function isValidLottoNumber(num) {
+  return Number.isInteger(num) && num >= 1 && num <= 45;
+}
+
+export function isValidLottoNumbersArray(arr) {
+  return Array.isArray(arr) && arr.every(isValidLottoNumber);
+}
+
+export function isDuplicateNumbersInArray(winningNumbers, bonusNumber) {
+  return winningNumbers.includes(bonusNumber);
+}
```
도메인 밖에 나와 있는 함수들은 가급적 도메인 지식을 모르도록 구현해주시면 좋겠습니다.
예를 들어 1, 45 같은 제한을 이 함수의 사용처(아마 도메인 쪽이겠죠?)서 주입하도록 구현해볼 수 있을 듯 해요.

Tanney-102 commented on src/domain/services/LottoService.js:
```diff
@@ -0,0 +1,54 @@
+
+export class LottoService {
+  constructor(totalAmount) {
+    this.lotto = new Lotto(totalAmount);
+    this.tickets = this.lotto.generateLottoTickets();
```
객체를 초기화 하는 시점과 티켓을 생성하는 시점이 같아도 괜찮을까요?
사용처에서 constructor를 호출할 때는 인스턴스를 만들고 각 프로퍼티를 기본값으로 초기화 하는 것 정도만 기대할 것 같은데 실제 핵심 기능 중 하나인 로또 티켓 발급을 이 시점에 하는 걸 에측하기 어려울 것 같아서요.

devpearlkim replied:
constructor에 너무 많은 로직을 넣었는데, 별 생각 없이 작성한 부분이 있었습니다.  
말씀해 주신 것처럼 constructor의 역할을 더 신중하게 고려하고 사용하도록 하겠습니다.  
핵심 로직이 constructor에 포함되어 있고, 티켓 발급 시점을 제어할 수 없는 문제가 있으므로  
별도 메서드로 분리하여 개선하겠습니다. 감사합니다!

Tanney-102 commented on src/domain/services/LottoService.js:
```diff
@@ -0,0 +1,54 @@
+  }
+
+  getLottoTickets() {
+    return this.tickets;
```
array의 레퍼런스를 그대로 반환해도 괜찮나요? 사용처에서 service.getLottoTickets().add(...) 이런식으로 변경을 주면 어떻게 될까요?

Tanney-102 commented on src/domain/services/LottoService.js:
```diff
@@ -0,0 +1,54 @@
+    return this.tickets.map((ticket) => this.lottoResult.calculateRank(ticket));
+  }
+
+  countWinningRanks() {
```
service는 게임 진행을 위해 각 모델간 소통만을 돕고 rankCount 같은건 도메인 쪽에서 연산하도록 하면 좋겠습니다.

Tanney-102 commented on src/domain/models/LottoStatistics.js:
```diff
@@ -0,0 +1,26 @@
+  constructor(rankCount, totalAmount) {
+    this.rankCount = rankCount;
+    this.totalAmount = totalAmount;
+    this.prizeMoney = {
+      1: 2000000000,
+      2: 30000000,
+      3: 1500000,
+      4: 50000,
+      5: 5000,
+    };
```
이건 멤버 변수 보다는 상수 같은 걸로 분리하면 좋겠습니다. 변수는 어디서든 변경될 여지가 있어서요

Tanney-102 commented on src/domain/models/Lotto.js:
```diff
@@ -0,0 +1,55 @@
+      tickets.add(JSON.stringify(newTicket));
+    }
+
+    // Set -> Array 변환
```
불필요한 주석입니다. 코드로 충분히 유추할 수 있는 정보에요

Tanney-102 commented on src/domain/models/Lotto.js:
```diff
@@ -0,0 +1,55 @@
+    }
+
+    // Set -> Array 변환
+    return Array.from(tickets).map((ticket) => JSON.parse(ticket));
```
ticket을 stringify했다가 다시 parse하는 이유가 있나요?

devpearlkim replied:
숫자 배열의 중복을 방지하려고 고민 없이 Set을 썼는데, 문자열로 변환하는 방식이 비효율적일 수도 있다는 생각은 못했습니다.
언급해주신 부분으로 찾아보니 Map을 활용하거나 join으로 문자열로 변환해서 저장하는게 더 적절한 방법일 것 같습니다.
감사합니다

Tanney-102 reviewed (changes requested):
안녕하세요 진주님! 리뷰어 오태은입니다.
요청 온 것을 놓쳐 리뷰가 많이 늦었습니다. 죄송해요🥲
작성해주신 코드 잘 봤습니다. 전반적으로 모듈 및 함수 분리여부, 네이밍 등 적절하게 잘 작성해주신 것 같습니다. 모듈 분리의 장점을 더 살리기 위해 몇가지 코멘트 남겨드렸으니 반영 후 다시 리뷰요청 부탁드립니다.

질문에 답변드립니다.

> LottoService의 setWinningNumbers() 호출 방식

객체가 변할 수 있는 상태를 가지고 있는 이유가 뭘까요~?
프로그램 안에서의 객체가 추상화하고 있는 현실 세계의 객체가 그 본질상 시간의 흐름에 따라 변하기 때문이 아닐까요?
프로그램이 실행됨에 따라 객체의 상태가 변하는 건 자연스러운 일입니다. 오히려 불필요한 변경이 있지는 않은지, 변경되지 말아야할 것이 변경될 여지가 있지는 않은지, 변경의 주도권을 누가 가지고 있는지 등에 대해 더 고민해보시면 좋겠습니다.

지금은 사이드 이펙트(user input) 이후 객체의 메서드를 통한 변경을 하고 있기 때문에 변경의 시점이나 방식 모두 문제 없어보입니다.

> LottoResult의 역할 분리

가지고 있는 멤버의 검증을 맡아서 하는 것도 꽤나 자연스럽니다. validation 쪽에 코멘트 드렸듯 그 검증의 주도권이 클래스 쪽으로 더 넘어올 수 있도록만 개선해주세요.

devpearlkim commented:
> 
> 객체가 변할 수 있는 상태를 가지고 있는 이유가 뭘까요~? 프로그램 안에서의 객체가 추상화하고 있는 현실 세계의 객체가 그 본질상 시간의 흐름에 따라 변하기 때문이 아닐까요? 프로그램이 실행됨에 따라 객체의 상태가 변하는 건 자연스러운 일입니다. 오히려 불필요한 변경이 있지는 않은지, 변경되지 말아야할 것이 변경될 여지가 있지는 않은지, 변경의 주도권을 누가 가지고 있는지 등에 대해 더 고민해보시면 좋겠습니다.

요구사항에 따라 간단한 로또 게임 로직을 작성했지만, 
실제 세계와 연결해서 생각해보지는 않았던 것 같습니다. 
말씀해주신 대로 다시 한번 객체의 상태와 메서드에 대해 고민하고 코드를 작성해야 겠다는 생각이 들었습니다.
말씀해주신 부분을 듣고 다른 자바 개발자분께도 여쭤봤는데, 제가 클래스, 상태, 메소드 개념을 제대로 모르고 있다는 것도 인지하게 되었습니다.

리뷰 감사합니다! 이 부분과 코멘트에 말씀해주신 부분을 확인하여 수정한 후 다시 리뷰 요청드리겠습니다.

감사합니다. 😊

devpearlkim committed:
fix: Avoid duplicate winning number assignment

devpearlkim committed:
test: Refine LottoResult boundary value validation checks

devpearlkim committed:
refactor: Return a copy of lotto tickets to prevent unintended modifications

devpearlkim committed:
refactor: Move prize money values to constants.js

devpearlkim committed:
Refactor: Use Map instead of Set to handle duplicate arrays efficiently

devpearlkim committed:
refactor: Remove checkResults from LottoService and move logic to LottoResult

- Removed checkResults() from LottoService and replaced it with getWinningRanks()
- Moved ranking calculation responsibility to LottoResult for better encapsulation
- Improved LottoService by reducing unnecessary logic and keeping it as a mediator

devpearlkim committed:
refactor: Seperate ticket generation fron LottoService constructor

devpearlkim committed:
refactor: Hide domain knowledge by passing LOTTO_NUMBER_RANGE as an object

devpearlkim committed:
refactor: Separate promptUser into a common module

Tanney-102 commented on src/validation.js:
```diff
@@ -0,0 +1,13 @@
+export function isValidLottoNumber(num, range) {
+  return Number.isInteger(num) && num >= range.min && num <= range.max;
+}
```
함수 내용이 더 일반적으로 바뀌었으니 이름에도 'Lotto' 같은 구체적인 내용이 들어갈 필요는 없겠죠!
`isInRange`? 같은 이름으로도 충분하지 않나 싶어요

Tanney-102 reviewed (approved):
안녕하세요 진주님! 빠르고 정확하게 리뷰내용을 잘 반영해주셨네요 ㅎㅎ
간단한 코멘트에도 충분히 고민 후 반영해주시는 게 느껴졌습니다.
사소한 코멘트 하나 더 남겼는데 이건 다음 스텝 진행하면서 반영해보시죠. 고생많으셨습니다!
